src/os_win/os_errno.c

```diff
@@ -63,14 +63,33 @@ int
 __wt_errno(void)
 {
 	/*
+	 * Check for 0:
+	 * It's easy to introduce a problem by calling the wrong error function,
+	 * for example, this function when the MSVC function set the C runtime
+	 * error value. Handle gracefully and always return an error.
+	 */
+	return (errno == 0 ? WT_ERROR : errno);
```

Comment:
I wouldn't object to adding the assertion back in HAVE_DIAGNOSTIC builds, if you want to do that.

I wouldn't try to clear `errno`; enough systems play games with the declaration of `errno` I doubt it's useful (or will even necessarily work).